File: backend/src/app/schemas/session.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from .character import CharacterSimple
from .campaign import Campaign # Import Campaign schema

# Base properties shared by all schemas
class SessionBase(BaseModel):
    title: str
    description: Optional[str] = None
    summary: Optional[str] = None
    date_time: Optional[datetime] = None # Allow session date/time to be optional

# Properties required for creation
class SessionCreate(SessionBase):
    campaign_id: int
    title: str # Ensure title is required on creation

# Properties required for update (all optional)
class SessionUpdate(SessionBase):
    title: Optional[str] = None # Override base to make it optional for update
    # campaign_id cannot be updated
    character_ids: Optional[List[int]] = None # Add list of character IDs for assignment

# Properties stored in DB but not returned to client directly
class SessionInDBBase(SessionBase):
    id: int
    campaign_id: int
    created_at: datetime
    updated_at: datetime # Should always have a value due to model defaults

    class Config:
        orm_mode = True # Use orm_mode for older Pydantic or from_attributes=True for v2

# Properties returned to client (can include relationships later if needed)
class Session(SessionInDBBase):
    characters: List[CharacterSimple] = [] # Add list of associated characters
    campaign: Optional[Campaign] = None # Add campaign relationship

# Properties stored in DB (internal use)
class SessionInDB(SessionInDBBase):
    pass 

File: frontend/src/types/session.ts
// import type { Optional } from './common'; // Assuming Optional exists or handle nullable types directly
import type { SessionSlot } from './session_slot'; // Added
import type { CharacterSimple } from './character'; // Import CharacterSimple as type
import type { Campaign } from './campaign'; // Import Campaign type

// Base properties shared by create/update/read
export interface SessionBase {
    title: string;
    description?: string | null;
    summary?: string | null;
    date_time?: string | null; // Use string for dates coming from JSON
    // availability_slots?: SessionSlot[]; // Moved to Session interface, not needed in base/create
}

// For creating a session
export interface SessionCreate extends SessionBase {
    campaign_id: number;
}

// For updating a session (all fields optional)
export interface SessionUpdate {
    title?: string;
    description?: string | null; // Use standard union type
    summary?: string | null;
    date_time?: string | null;
    character_ids?: number[]; // Add character IDs for update
}

// Session object received from API
export interface Session extends SessionBase {
    id: number;
    campaign_id: number;
    created_at: string; // ISO Date string - Belongs here
    updated_at: string; // ISO Date string - Belongs here
    availability_slots?: SessionSlot[]; // Optional relationship data
    characters: CharacterSimple[]; // Add associated characters
    campaign?: Campaign; // Add optional campaign relationship
    // Add relationships here later if needed (e.g., characters, entries)
} 